util/topsql/reporter/datasink.go

```diff
@@ -26,3 +36,106 @@ type DataSink interface {
 	OnReporterClosing()
 }
+
+// DataSinkRegisterer is for registering DataSink
+type DataSinkRegisterer interface {
+	Register(dataSink DataSink) error
+	Deregister(dataSink DataSink)
+}
+
+// ReportData contains data that reporter sends to the agent.
+type ReportData struct {
+	// DataRecords contains the topN records []tipb.TopSQLRecord and the `others`
+	// record which aggregation all []tipb.TopSQLRecord that is out of Top N.
+	DataRecords []tipb.TopSQLRecord
+	SQLMetas    []tipb.SQLMeta
+	PlanMetas   []tipb.PlanMeta
+}
+
+func (d *ReportData) hasData() bool {
+	return len(d.DataRecords) != 0 || len(d.SQLMetas) != 0 || len(d.PlanMetas) != 0
+}
+
+var _ DataSink = &DefaultDataSinkRegisterer{}
+var _ DataSinkRegisterer = &DefaultDataSinkRegisterer{}
+
+// DefaultDataSinkRegisterer implements DataSinkRegisterer.
+type DefaultDataSinkRegisterer struct {
+	sync.Mutex
+	ctx       context.Context
+	dataSinks map[DataSink]struct{}
+}
+
+// NewDefaultDataSinkRegisterer creates a new DefaultDataSinkRegisterer which implements DataSinkRegisterer.
+func NewDefaultDataSinkRegisterer(ctx context.Context) DefaultDataSinkRegisterer {
+	return DefaultDataSinkRegisterer{
+		ctx:       ctx,
+		dataSinks: make(map[DataSink]struct{}, 10),
+	}
+}
+
+// Register implements DataSinkRegisterer.
+func (r *DefaultDataSinkRegisterer) Register(dataSink DataSink) error {
+	r.Lock()
+	defer r.Unlock()
+
+	select {
+	case <-r.ctx.Done():
+		return errors.New("DefaultDataSinkRegisterer closed")
+	default:
+		if len(r.dataSinks) >= 10 {
+			return errors.New("too many datasinks")
+		}
+		r.dataSinks[dataSink] = struct{}{}
+		if len(r.dataSinks) > 0 {
+			variable.TopSQLVariable.Enable.Store(true)
+		}
+		return nil
+	}
+}
+
+// Deregister implements DataSinkRegisterer.
+func (r *DefaultDataSinkRegisterer) Deregister(dataSink DataSink) {
+	r.Lock()
+	defer r.Unlock()
+
+	select {
+	case <-r.ctx.Done():
+	default:
+		delete(r.dataSinks, dataSink)
+		if len(r.dataSinks) == 0 {
+			variable.TopSQLVariable.Enable.Store(false)
+		}
+	}
+}
+
+// TrySend implements DataSink.
+//
+// TrySend sends ReportData to all internal registered DataSinks.
+func (r *DefaultDataSinkRegisterer) TrySend(data *ReportData, deadline time.Time) error {
+	r.Lock()
+	dataSinks := make([]DataSink, 0, len(r.dataSinks))
+	for ds := range r.dataSinks {
+		dataSinks = append(dataSinks, ds)
+	}
+	r.Unlock()
+	for _, ds := range dataSinks {
+		if err := ds.TrySend(data, deadline); err != nil {
+			logutil.BgLogger().Warn("[top-sql] failed to send data to datasink", zap.Error(err))
+		}
+	}
+	return nil
+}
+
+// OnReporterClosing implements DataSink.
+//
+// OnReporterClosing calls the OnReporterClosing method of all internally registered DataSinks.
+func (r *DefaultDataSinkRegisterer) OnReporterClosing() {
+	var m map[DataSink]struct{}
+	r.Lock()
+	m, r.dataSinks = r.dataSinks, make(map[DataSink]struct{})
+	r.Unlock()
+	for d := range m {
+		d.OnReporterClosing()
+	}
+}
```

Comment:
Why `DefaultDataSinkRegisterer` implements `DataSink`?

Reply:
Implementing private helper functions `trySend` and `onReporterClosing` for `RemoteTopSQLReporter` is better for me.

Reply:
`DefaultDataSinkRegisterer` is embedded by `RemoteTopSQLReporter` as an implementation of `DataSinkRegisterer`, and is also called by `RemoteTopSQLReporter` as an implementation of `DataSink`. When used as the implementation of `DataSink`, any call made to `DefaultDataSinkRegisterer` is equivalent to making separate calls to the internally registered `DataSink`.

Reply:
> Implementing private helper functions `trySend` and `onReporterClosing` for `RemoteTopSQLReporter` is better for me.

Of course, maybe we should not expose the public method of `DataSink` for `RemoteTopSQLReporter`.